### PR TITLE
[1.9.x] Intraprocess delivery in StatelessWriter [6806]

### PR DESCRIPTION
**Pull request by adolfomarver**

### CODE REVIEW

File: include/fastrtps/rtps/writer/ReaderLocator.h [hunk 1]
```diff
@@ -64,6 +65,21 @@ class ReaderLocator : public RTPSMessageSenderInterface
...
+        {
+            return is_local_reader_;
+        }
+
+        RTPSReader* local_reader()
+        {
+            return local_reader_;
+        }
+
+        void set_local_reader(RTPSReader* local_reader)
```
Comment (MiguelCompany) [inline include/fastrtps/rtps/writer/ReaderLocator.h]:
```suggestion
        void local_reader(
                RTPSReader* local_reader)
```

File: include/fastrtps/rtps/writer/StatelessWriter.h [hunk 2]
```diff
@@ -143,6 +143,8 @@ class StatelessWriter : public RTPSWriter
 
     void update_reader_info(bool create_sender_resources);
 
+    bool intraprocess_delivery(CacheChange_t* change, ReaderLocator& reader_locator);
```
Comment (MiguelCompany) [inline include/fastrtps/rtps/writer/StatelessWriter.h]:
```suggestion
    bool intraprocess_delivery(
            CacheChange_t* change,
            ReaderLocator& reader_locator);
```

File: src/cpp/rtps/writer/ReaderLocator.cpp [hunk 3]
```diff
@@ -36,6 +36,7 @@ ReaderLocator::ReaderLocator(
     : owner_(owner)
     , locator_info_(max_unicast_locators, max_multicast_locators)
     , expects_inline_qos_(false)
+    , is_local_reader_(false)
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/ReaderLocator.cpp]: Field `local_reader_` should be initialized to `nullptr`

File: src/cpp/rtps/writer/ReaderProxy.cpp [hunk 4]
```diff
@@ -81,7 +81,8 @@ void ReaderProxy::start(const ReaderProxyData& reader_attributes)
         reader_attributes.guid(),
         reader_attributes.remote_locators().unicast,
         reader_attributes.remote_locators().multicast,
-        reader_attributes.m_expectsInlineQos);
+        reader_attributes.m_expectsInlineQos,
+        /*reader_attributes.guid().is_on_same_process_as(writer_->getGuid()*/false);
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/ReaderProxy.cpp]: Add a TODO explaining why this is being done

File: src/cpp/rtps/writer/StatelessWriter.cpp [hunk 5]
```diff
@@ -34,6 +34,7 @@
 #include <vector>
 
 #include <fastrtps/log/Log.h>
+#include "../RTPSDomainImpl.hpp"
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/StatelessWriter.cpp]:
```suggestion
#include "rtps/RTPSDomainImpl.hpp"
```

File: src/cpp/rtps/writer/StatelessWriter.cpp [hunk 6]
```diff
@@ -207,14 +207,48 @@ void StatelessWriter::unsent_change_added_to_history(
...
         }
     }
+
+    if (no_readers_to_add_change)
+    {
+        logInfo(RTPS_WRITER, "No reader to add change.");
+    }
+}
+
+bool StatelessWriter::intraprocess_delivery(CacheChange_t* change, ReaderLocator& reader_locator)
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/StatelessWriter.cpp]: Each parameter on a new line

File: src/cpp/rtps/writer/StatelessWriter.cpp [hunk 7]
```diff
@@ -207,14 +207,48 @@ void StatelessWriter::unsent_change_added_to_history(
...
+{
+    RTPSReader* reader = reader_locator.local_reader();
+    
+    if (!reader)
+    {
+        if((reader = RTPSDomainImpl::find_local_reader(reader_locator.remote_guid())))
+        {
+            reader_locator.set_local_reader(reader);
```
Comment (MiguelCompany) [inline src/cpp/rtps/writer/StatelessWriter.cpp]:
```suggestion
            reader_locator.local_reader(reader);
```

File: src/cpp/rtps/writer/StatelessWriter.cpp (hunk 5 above)
Comment (richiware) [reply src/cpp/rtps/writer/StatelessWriter.cpp]: `"rtps/RTPSDomainImpl.hpp"` will not be found using relative paths. The suggestion is wrong.

**Pull request closed without merging**